lib/spack/spack/ci.py

```diff
@@ -542,11 +540,21 @@ def get_spec_filter_list(env, affected_pkgs, dependencies=True, dependents=True)
     for pkg in affected_pkgs:
         env_matches = [s for s in all_concrete_specs if s.name == pkg]
         for match in env_matches:
-            affected_specs.add(match)
-            if dependencies:
-                affected_specs.update(match.traverse(direction="children", root=False))
-            if dependents:
-                affected_specs.update(match.traverse(direction="parents", root=False))
+            visited = set()
+            affected_specs.update(
+                match.traverse(
+                    direction="children", root=True, visited=visited, key=lambda s: s.dag_hash()
+                )
+            )
+            for dependent_spec in match.traverse(direction="parents", root=False):
```

Comment:
```suggestion
            for dependent_spec in match.traverse(direction="parents", root=True):
```

Reply:
Ah, sorry, I missed that second `root=True`. Can you still simplify to

```python
visited = set()
dag_hash = lambda s: s.dag_hash()
for parent in match.traverse(direction="parents", key=dag_hash):
    affected_specs.update(parent.traverse(direction="children", visited=visited, key=dag_hash)
```

it's much more succinct

Reply:
and in fact you can drop the outer loop with `env_matches = [s for s in all_concrete_specs if s.name in affected_pkgs]`